src/ray/core_worker/core_worker.cc

```diff
@@ -2759,13 +2759,11 @@ void CoreWorker::HandleExit(const rpc::ExitRequest &request, rpc::ExitReply *rep
                             rpc::SendReplyCallback send_reply_callback) {
   bool own_objects = reference_counter_->OwnObjects();
   // Fail the request if it owns any object.
-  if (own_objects) {
-    reply->set_success(false);
-  } else {
-    reply->set_success(true);
+  reply->set_success(!own_objects);
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+  if (!own_objects) {
```

Comment:
We need to send the reply before exiting.

Reply:
It is actually fine because Exit is posted to the event loop. But the current structure sounds cleaner anyway.

Reply:
`send_reply_callback` probably also posts to the event loop, right? @iycheng, could you add a comment about the ordering?

Reply:
I'm not quite sure about this actually. I checked the code, it goes to task execution pool, but the grpc request goes to io pool.

I feel there still might be some bugs here, like worker exited before sending the request. I saw logs with info like: Error: worker failure with  INTEND_EXITING. I kind of feeling that if it exited in the normal way, it shouldn't complain about this.

Reply:
@stephanie-wang that's correct. I feel we need to make it sync here.

Reply:
How about putting Exit into the callback?

Reply:
I think it would be better to just handle this on the raylet side and not try to get the ordering exactly right. The raylet needs to be robust to the worker failing before successfully sending the reply anyway.

Reply:
It looks like it doesn't post to the io service when send_reply_callback is called. https://github.com/ray-project/ray/blob/9b3ea7c32b8a85ac35c01724689e67e9c7e35893/src/ray/rpc/server_call.h#L164

Reply:
@stephanie-wang I agreed that raylet should be robust. I already added the logic there to remove the worker.

But here, I think it's not worker failure. It's about intending exiting which we can make it safe.